File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/PmdMeasurementType.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd

enum class PmdMeasurementType(val numVal: UByte) {
    ECG(0u),
    PPG(1u),
    ACC(2u),
    PPI(3u),
    GYRO(5u),
    MAGNETOMETER(6u),
    SKIN_TEMP(7u),
    SDK_MODE(9u),
    LOCATION(10u),
    PRESSURE(11u),
    TEMPERATURE(12u),
    OFFLINE_RECORDING(13u),
    OFFLINE_HR(14u),
    OFFLINE_TEMP(15u),
    UNKNOWN_TYPE(0x3fu);

    fun isDataType(): Boolean {
        return when (this) {
            SDK_MODE,
            OFFLINE_RECORDING,
            UNKNOWN_TYPE -> false
            else -> true
        }
    }

    internal companion object {
        private const val MEASUREMENT_TYPE_BIT_MASK: UByte = 0x3Fu

        fun fromId(id: Byte): PmdMeasurementType {
            for (type in values()) {
                if (type.numVal == (id.toUByte() and MEASUREMENT_TYPE_BIT_MASK)) {
                    return type
                }
            }
            return UNKNOWN_TYPE
        }

        fun fromByteArray(data: ByteArray): Set<PmdMeasurementType> {

            val measurementTypes: MutableSet<PmdMeasurementType> = mutableSetOf()
            if ((data[1].toUInt() and 0x01u) != 0u) measurementTypes.add(ECG)
            if ((data[1].toUInt() and 0x02u) != 0u) measurementTypes.add(PPG)
            if ((data[1].toUInt() and 0x04u) != 0u) measurementTypes.add(ACC)
            if ((data[1].toUInt() and 0x08u) != 0u) measurementTypes.add(PPI)
            if ((data[1].toUInt() and 0x20u) != 0u) measurementTypes.add(GYRO)
            if ((data[1].toUInt() and 0x40u) != 0u) measurementTypes.add(MAGNETOMETER)
            if ((data[1].toUInt() and 0x80u) != 0u) measurementTypes.add(SKIN_TEMP)
            if ((data[2].toUInt() and 0x02u) != 0u) measurementTypes.add(SDK_MODE)
            if ((data[2].toUInt() and 0x04u) != 0u) measurementTypes.add(LOCATION)
            if ((data[2].toUInt() and 0x08u) != 0u) measurementTypes.add(PRESSURE)
            if ((data[2].toUInt() and 0x10u) != 0u) measurementTypes.add(TEMPERATURE)
            if ((data[2].toUInt() and 0x20u) != 0u) measurementTypes.add(OFFLINE_RECORDING)
            if ((data[2].toUInt() and 0x40u) != 0u) measurementTypes.add(OFFLINE_HR)
            return measurementTypes
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/PmdMeasurementType.swift
import Foundation

public enum PmdMeasurementType: UInt8, CaseIterable {
    case ecg = 0
    case ppg = 1
    case acc = 2
    case ppi = 3
    case gyro = 5
    case mgn = 6
    case sdkMode = 9
    case location = 10
    case pressure = 11
    case temperature = 12
    case offline_recording = 13
    case offline_hr = 14
    case unknown_type = 0x3f
    
    private static let MEASUREMENT_BIT_MASK: UInt8 = 0x3F
    
    func isDataType() -> Bool {
        switch(self) {
        case .sdkMode, .offline_recording, .unknown_type:
            return false
        default :
            return true
        }
    }
    
    static func fromId(id: UInt8) -> PmdMeasurementType {
        for type in PmdMeasurementType.allCases {
            if (type.rawValue == (id & PmdMeasurementType.MEASUREMENT_BIT_MASK)) {
                return type
            }
        }
        return .unknown_type
    }
    
    static func fromByteArray(_ data: Data) -> Set<PmdMeasurementType> {
        var measurementTypes:Set<PmdMeasurementType> = []
        if (data[1] & 0x01 != 0) {
            measurementTypes.insert(PmdMeasurementType.ecg)
        }
        if (data[1] & 0x02 != 0) {
            measurementTypes.insert(PmdMeasurementType.ppg)
        }
        if (data[1] & 0x04 != 0) {
            measurementTypes.insert(PmdMeasurementType.acc)
        }
        if (data[1] & 0x08 != 0) {
            measurementTypes.insert(PmdMeasurementType.ppi)
        }
        if (data[1] & 0x20 != 0) {
            measurementTypes.insert(PmdMeasurementType.gyro)
        }
        if (data[1] & 0x40 != 0) {
            measurementTypes.insert(PmdMeasurementType.mgn)
        }
        if (data[2] & 0x04 != 0) {
            measurementTypes.insert(PmdMeasurementType.location)
        }
        if (data[2] & 0x08 != 0) {
            measurementTypes.insert(PmdMeasurementType.pressure)
        }
        if (data[2] & 0x10 != 0) {
            measurementTypes.insert(PmdMeasurementType.temperature)
        }
        if (data[2] & 0x40 != 0) {
            measurementTypes.insert(PmdMeasurementType.offline_hr)
        }
        
        if (data[2] & 0x02 != 0) {
            measurementTypes.insert(PmdMeasurementType.sdkMode)
        }
        if (data[2] & 0x20 != 0) {
            measurementTypes.insert(PmdMeasurementType.offline_recording)
        }
        
        return measurementTypes
    }
}
